### ISSUE TITLE
Recommended way to read the Request.Body in an ActionFilter

### ISSUE THREAD

**Issue by damienbod** [aspnet/Mvc#5260]
What's the recommended way to read the request body inside an Actionfilter

The Body property in the ActionExecutingContext cannot be read.

Thanks Damien

**Comment by davidfowl** [comment]
What are you trying to do with the body?

**Comment by damienbod** [comment]
I want to validate the model

**Comment by damienbod** [comment]
Are you trying to prevent developers from implementing validation logic in Filters? After reading the docs, this is not one of the use cases for ActionFilters.

Greetings Damien

**Comment by rynowak** [comment]
If you must do this inside an action filter then you'll need to buffer the body (new up a memory stream and copy to it, replace the body). ModelBinding runs before action filters so if you have form data or a `[FromBody]` parameter, we've already read it.

If you can tell us more about what you need to accomplish we can suggest alternatives

**Comment by damienbod** [comment]
Hi @rynowak thanks for your answer. I don't have to validate in the ActionFilter, I just though this would be a good place to put the validation logic before I call the controller method. I use the FromBody in the Controller class to get the data. Can I access this in an ActionFilter? 

If not I can always do the validation from the controller class inside the method.

Thanks Damien

**Comment by rynowak** [comment]
Yes, if you're inside an action filter then `context.ActionArguments` will contain all of the model objects we created. So if you have:

```
public IActionResult Edit(int id, [FromBody] Widget widget) { }
```

Then `context.ActionArguments["widget"]` will return the `Widget` object. If you're trying to do this in a generic way, look at `context.ActionDescriptor.Parameters` - this will contain all of the parameter definitions and metadata.

**Comment by damienbod** [comment]
Cool thanks, works perfect. Thanks a million for the tip.

Greetings Damien

**Comment by damienbod** [comment]
Created a blog about this. thanks
https://damienbod.com/2016/09/09/asp-net-core-action-arguments-validation-using-an-actionfilter/

**Comment by nhwilly** [comment]
Interestingly, if you try to update the model (as I am doing with inbound open id connect claims), the validation has already occurred.  You have to force it to validate again.  

I tried using a resource filter so I could catch it first, but I couldn't figure out how to determine the object type (which would tell me I needed to insert the data into the model).  I can't always assume there is just one model coming in, right?  One of the incoming models might not have the properties I need to update.  

So I loop through, find the models that need updating, make the updates and then force validation again.

Doesn't sound very efficient, but I can't figure another way to do it.  Any suggestions/pointers would be great.

Thanks for the articles and the answers here.  Huge help.

**Comment by rynowak** [comment]
@nhwilly It would help if you could open a new issue and explain what you're trying to do from the beginning.

**Comment by nhwilly** [comment]
Sorry. Thanks for asking.  

During a user's login, I issue a JWT token for authentication containing claims for their user id and their account id.  That token is presented for each call to my api.

For *almost* every document I store, I'll need those values to be present in my final entities.  So I thought I would use an action filter to centralize that logic. 

To do that, I added two fields to my data transfer pocos and wanted the filter to populate them so by the time my controller gets the incoming model argument, it already contains the values.  The fields have been marked as required in the poco, so validation will pick them up, if somehow they are not present. 

Automapper will then move them to my full entities.

But the ModelState is already set to invalid by the time the action filter gets it, since at the very least, those two values haven't been filled and binding is run prior to my action filter being executed. Filling in those values may actually make the model valid.  In fact, most times it will.

In my code, if the incoming model inherits frrom "Accountable", I populate the fields, clear the ModelState errors and revalidate.

It works, but maybe I've gone about it all wrong.

Maybe I should remove the required attributes on those fields, let model binding do it's thing for simple attributes and use FluentValidation after Automapper gets finished to catch everything else, including account id and user id.

I expect to have to use Fluent anyway for more complex validation down the road.

Is there a better way?

Thanks again for responding.

Disclaimer: It's shortly after 4am here, the coyotes woke me up and I decided to check my email. In case this sounds obviously foolish. 🙂
